### PR TITLE
[BEAM-53] PubsubUnboundedSource

### PR DESCRIPTION
**Pull request by mshields822**
This adds a PubsubUnboundedSource with supporting utilities and tests.

Not yet wired into PubsubIO.Read, I'll do that in next PR so we can switch to source and sink atomically.

R: @dhalperi

### CODE REVIEW

**Comment by mshields822** [comment]
Forgot to mention: I still need to add unit tests for the checkpointing and ack timeouts. Will do that concurrently with the main review. Thanks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 1]
```diff
@@ -0,0 +1,1195 @@
...
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.util.MovingFunction;
+import org.apache.beam.sdk.util.PubsubClient;
+import org.apache.beam.sdk.util.PubsubClient.PubsubClientFactory;
+import org.apache.beam.sdk.util.PubsubClient.SubscriptionPath;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+
+import com.google.api.client.util.Clock;
+import com.google.api.client.util.Preconditions;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: bad imports here. I suspect maven will complain if you compile this.

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: nope, this means we need to up the checkstyle checks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 2]
```diff
@@ -0,0 +1,1195 @@
...
+ * delivers the oldest (in Pubsub processing time) available message at least once a minute,
+ * and that custom timestamps are 'mostly' monotonic with Pubsub processing time. Unfortunately
+ * both of those assumptions are fragile. Thus the estimated watermark may get ahead of
+ * the 'true' watermark and cause some messages to be late.
+ * <li>Checkpoints are used both to ACK received messages back to Pubsub (so that they may
+ * be retired on the Pubsub end), and to NACK already consumed messages should a checkpoint
+ * need to be restored (so that Pubsub will resend those messages promptly).
+ * <li>The backlog is determined by each reader using the messages which have been pulled from
+ * Pubsub but not yet consumed downstream. The backlog does not take account of any messages queued
+ * by Pubsub for the subscription. Unfortunately there is currently no public API to determine
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: drop the word "public" here?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 3]
```diff
@@ -0,0 +1,1195 @@
...
+ * the size of the Pubsub queue's backlog.
+ * <li>The subscription must already exist.
+ * <li>The subscription timeout is read whenever a reader is started. However it is not
+ * checked thereafter despite the timeout being user-changeable on-the-fly.
+ * <li>We log vital stats every 30 seconds.
+ * <li>Though some background threads may be used by the underlying transport all Pubsub calls
+ * are blocking. We rely on the underlying runner to allow multiple
+ * {@link UnboundedSource.UnboundedReader} instance to execute concurrently and thus hide latency.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: instances.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 4]
```diff
@@ -0,0 +1,1195 @@
...
+        @Override
+        public long apply(long left, long right) {
+          return left + right;
+        }
+
+        @Override
+        public long identity() {
+          return 0;
+        }
+      };
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: This one seems to exist already: https://github.com/apache/incubator-beam/blob/96765f19b1bd8149240cd77eb7cf7fb636e477e4/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sum.java#L149

I'm not sure why the others don't, honestly. Should we just add them?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 5]
```diff
@@ -0,0 +1,1195 @@
...
+
+    /**
+     * BLOCKING
+     * All messages which have been passed downstream have now been durably committed.
+     * We can ACK them upstream.
+     * CAUTION: This may never be called.
+     */
+    @Override
+    public void finalizeCheckpoint() throws IOException {
+      Preconditions.checkState(reader != null && safeToAckIds != null,
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: static import

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 6]
```diff
@@ -0,0 +1,1195 @@
...
+    public PubsubReader<T> createReader(
+        PipelineOptions options,
+        @Nullable PubsubCheckpoint<T> checkpoint) {
+      PubsubReader<T> reader;
+      try {
+        reader = new PubsubReader<>(options.as(PubsubOptions.class), this);
+      } catch (GeneralSecurityException | IOException e) {
+        throw new RuntimeException("Unable to subscribe to " + outer.subscription + ": ", e);
+      }
+      if (checkpoint != null) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: move this logic to `start()`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    Arrays.fill(numSamples, 0);
+    currentMsSinceEpoch = -1;
+    currentIndex = -1;
+  }
+
+  /**
+   * Flush stale values.
+   */
+  private void flush(long nowMsSinceEpoch) {
+    Preconditions.checkArgument(nowMsSinceEpoch >= 0);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java]: static import

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java]: error messages

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    flush(nowMsSinceEpoch);
+    buckets[currentIndex] = function.apply(buckets[currentIndex], value);
+    numSamples[currentIndex]++;
+  }
+
+  /**
+   * Return the minimum/maximum/sum of all retained values within {@link #samplePeriodMs}
+   * of {@code nowMsSinceEpoch}.
+   */
+  public long get(long nowMsSinceEpoch) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java]: if all these times are really `now`, why not just allow passing a clock in the constructor?

Otherwise, rename the vars?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java [hunk 9]
```diff
@@ -118,18 +125,15 @@ public String getKind() {
...
   private Map<String, Long> ackDeadline;
 
-  /**
-   * Current time.
-   */
-  private long nowMsSinceEpoch;
-
   @VisibleForTesting
   PubsubTestClient(
+      @Nullable Clock clock,
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java]: not sure it's actually nullable? I don't see anywhere you check for nullity.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 10]
```diff
@@ -0,0 +1,1195 @@
...
+      pubsubClient.modifyAckDeadline(outer.outer.subscription, ackIds, extensionSec);
+      numExtendedDeadlines.add(nowMsSinceEpoch, ackIds.size());
+    }
+
+    /**
+     * Messages which have been ACKed (via the checkpoint finalize) are no longer in flight.
+     * This is only used for flow control and stats.
+     */
+    private void retire() throws IOException {
+      long nowMsSinceEpoch = outer.outer.clock.currentTimeMillis();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: what is this `outer.outer.clock` pattern. Can it be simplified into something like `getClock()`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 11]
```diff
@@ -0,0 +1,1195 @@
...
+
+          // Extend the ACK for this message.
+          toBeExtended.add(entry.getKey());
+          if (toBeExtended.size() >= ACK_BATCH_SIZE) {
+            // Enough for one batch.
+            break;
+          }
+        }
+
+        if (toBeExtended.isEmpty() && toBeExpired.isEmpty()) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: this block seems of little value given that this will clearly happen in fallthrough. Strong need?

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: oh, this is just the way you wrote the while loop. a little ugly, but fine.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 12]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
...
+     * {@link UnboundedReader}.
+     * <li>It is possible for multiple checkpoints created from the same {@link UnboundedReader}
+     * to be 'in-flight' (ie not yet finalized) simultaneously.
+     * <li>It is possible for a checkpoint to never be finalized.
+     * <li>If this call throws an exception then the entire checkpoint will be abandoned and the
+     * reader restarted from an earlier, successfully-finalized checkpoint.
+     * <li>The system ensures that if a checkpoint fails for any reason then no later checkpoint
+     * is allowed to be finalized without the reader first being restarted.
+     * <li>It is not safe to assume the {@link UnboundedReader} from which this checkpoint was
+     * created still exists at the time this method is called.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: This change to general API (or the documentation thereof) should be in its own PR and sent to the dev list. This is a good set of things to clarify though! I'd also like the javadoc changes to be iterated on by folks like 

@kennknowles @lukecwik @dpmills @aljoscha @mxm @amitsela

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 13]
```diff
@@ -0,0 +1,1195 @@
...
+      }
+
+      String oldestInFlight = "no";
+      String oldestAckId = Iterables.getFirst(inFlight.keySet(), null);
+      if (oldestAckId != null) {
+        oldestInFlight =
+            (nowMsSinceEpoch - inFlight.get(oldestAckId).requestTimeMsSinceEpoch) + "ms";
+      }
+
+      LOG.warn("Pubsub {} has "
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: `info`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 12 above)
Comment (aljoscha) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Oh, we're not even calling that in the Flink runner. I think it should be possible to add it in the `UnboundedSourceWrapper`, though.

**Comment by dhalperi** [comment]
In general, looks pretty good. Some minor comments here and there.

The only major one is factoring out the UnboundedSource doc change separately and to a wider audience.

Overall feedback:
- large CL, hard to review confidently.
- but I know you've been running pretty stressful integration tests locally, so I'm giving you benefit of the doubt there.
- unit tests are thin, looking forward to seeing the ones you referenced.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 1 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: Fixed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 2 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 3 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 4 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: https://issues.apache.org/jira/browse/BEAM-285

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 5 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 6 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: This is our last chance to work with the checkpoint.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java (hunk 7 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java]: done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java (hunk 8 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java]: Don't want to pass Clock since I snapshot 'now' per logical unit of work.
Don't want to rename since I like to spell out exact units.
(And don't want to use Instant since it is surprisingly expensive.)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java (hunk 9 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java]: Fixed. (Basically, only needed for subscribe tests, not publish tests).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 10 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 11 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: yeah, bit ugly, sorry.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 12 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Agree - I've been letting the comments mature as I went. Hoisting...

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java (hunk 13 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: done

**Comment by mshields822** [comment]
PTAL
Some cascading changes once I dug deeper into checkpointing unit tests.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java [hunk 14]
```diff
@@ -184,4 +189,72 @@ public void timeoutAckExtensions() throws IOException {
...
+      // Preserve timestamp.
+      assertEquals(new Instant(messageNumToTimestamp(messageNum)), reader.getCurrentTimestamp());
+      // Preserve record id.
+      String recid = String.format("recordid_%d", messageNum);
+      assertArrayEquals(recid.getBytes(), reader.getCurrentRecordId());
+
+      if (i % 1000 == 999) {
+        // Estimated watermark can never get ahead of actual outstanding messages.
+        long watermark = reader.getWatermark().getMillis();
+        System.out.printf("**** watermark %d ****\n", watermark);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java]: remove?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java [hunk 15]
```diff
@@ -184,4 +189,72 @@ public void timeoutAckExtensions() throws IOException {
...
+        long minOutstandingTimestamp = Long.MAX_VALUE;
+        for (Integer outstandingMessageNum : dataToMessageNum.values()) {
+          minOutstandingTimestamp =
+              Math.min(minOutstandingTimestamp, messageNumToTimestamp(outstandingMessageNum));
+        }
+        assertTrue(watermark <= minOutstandingTimestamp);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java]: can you use Matchers to get better failure messages? Something like

```java
assertThat(watermark, lessThanOrEqualTo(minOutstandingTimestamp));
```

**Comment by dhalperi** [comment]
LGTM. Jenkins is being flaky. Can you take one more pass over the test assert messages so that if they break we have a hope of fixing them?

Let me know when you'd like me to merge.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java (hunk 14 above)
Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java]: done, sorry.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java (hunk 15 above)
Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSourceTest.java]: sure - that was the only hamcrest dep so I was holding out...

**Comment by mshields822** [comment]
PTAL
Fixed the problems with serialization + concurrent readers on PubsubTestClient Thomas noticed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java [hunk 16]
```diff
@@ -37,249 +39,307 @@
...
-    this.expectedTopic = expectedTopic;
-    this.remainingExpectedOutgoingMessages = expectedOutgoingMessages;
-    this.clock = clock;
-    this.expectedSubscription = expectedSubscription;
-    this.ackTimeoutSec = ackTimeoutSec;
-    this.remainingPendingIncomingMessages = expectedIncomingMessages;
-    this.pendingAckIncomingMessages = pendingAckIncomingMessages;
-    this.ackDeadline = ackDeadline;
+  private boolean inPullMode() {
+    checkState(STATE.isActive, "No test is active");
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java]: do these need synchrony?

Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubTestClient.java]: always called under STATE lock
